core/src/main/java/org/elasticsearch/http/HttpServer.java

```diff
@@ -98,7 +100,12 @@ public void dispatchRequest(HttpRequest request, HttpChannel channel, ThreadCont
             handleFavicon(request, channel);
             return;
         }
-        restController.dispatchRequest(request, channel, threadContext);
+        try {
+            transport.addBytes(request);
+            restController.dispatchRequest(request, new ResourceHandlingHttpChannel(channel, transport), threadContext);
+        } catch (CircuitBreakingException ex) {
```

Comment:
This is very subtle, but I think we cannot do this. `HttpServer` will send the error response on the raw Netty channel wherea `RestController` will send the error response on `ResourceHandlingHttpChannel`. In the former case no bytes will be freed (as no bytes have been reserved) and in the latter case we ensure that all reserved bytes are freed again. 

But I can still move all exception handling logic into `HttpServer`. I just have to keep track via which channel I send an error response. Something along these lines should do it:

```java
HttpChannel responseChannel = channel;
try {
  transport.addBytes(request);
  responseChannel = new ResourceHandlingHttpChannel(channel, transport);
  restController.dispatchRequest(request, responseChannel, threadContext);
} catch (Throwable t) {
  restController.sendErrorResponse(request, responseChannel, t);
}
```